### DISCUSSION TITLE
Error --via-ir

### DISCUSSION THREAD

**Discussion by Abhishek-2416** [Cyfrin/foundry-full-course-cu#1486] [Q&A] [answered]
my terminal throws an error 
Error: 
Compiler run failed:
Error: Compiler error (/solidity/libsolidity/codegen/CompilerUtils.cpp:1430):Stack too deep. Try compiling with `--via-ir` (cli) or the equivalent `viaIR: true` (standard JSON) while enabling the optimizer. Otherwise, try removing local variables.

i tried running forge coverage --via-ir still shows the same error , any other solution for this

**Comment by EngrPips** [comment]
Hello @Abhishek-2416, Stack too deep errors suck. They are almost always cause by limitations of the EVM and all of the data it needs to handle. This is a link https://ethereum.stackexchange.com/questions/12736/stack-too-deep-exception to a similar issues in stack exchange, please take a look maybe it would help resolve your issue.

**Reply by Abhishek-2416** [reply]
Is there any other way to solve this ?

**Reply by EngrPips** [reply]
Okay, i found this too https://ethereum.stackexchange.com/questions/6061/error-while-compiling-stack-too-deep on ethereum stack exchange and it really explains what could cause the `stack too deep` error and the possible mitigation. i think this would help

**Reply by Abhishek-2416** [reply]
This was the thing i tried 
before this i was getting the error even when i was trying to do forge test
Now it shows this error just for forge coverage for some reason

**Reply by Abbasjafri-syed** [reply, marked as answer]
hey,
 was getting this error also...finds that Stack too deep error is thrown by the Solidity compiler when any function have lots of local variables, have complex logical expressions or have a deep function stack..

The main issue is due to using lot of local variable in setup of raffle test which can be mitigated by dividing it into 2 parts like below:
```

    function setUp() public {
        // setting function for test
        deployRaffle deploy_raffle = new deployRaffle(); // deploying raffle contract through script
        (raffle, helper_mode) = deploy_raffle.run(); // assigning contract saved to their type

        (chainlink_Feed, price, coordinator_vrf, hash,,,,) = helper_mode.active_Config(); //caching value
        (,,,, call_Limit, confirmations, num_Values, subscriptionId) = helper_mode.active_Config(); // breaking to avoid stack too deep error
        vm.deal(funder, init_Value); // funding user with amount
    }
```
try breaking into 2 parts it will solve your compiler issue.
i have added my own params as per need but the main thing is break it into 2 parts

**Reply by Abhishek-2416** [reply]
Tried it just now 
doesnt seem to be working mate

**Reply by Abbasjafri-syed** [reply]
can you share the code

**Reply by Abhishek-2416** [reply]
```
function setUp() external {
        DeployRaffle deployer = new DeployRaffle();
        (raffle, helperConfig) = deployer.run();

        (entranceFee, interval, vrfCoordinator, gaslane,,,,) = helperConfig.activeNetworkConfig();
        (,,,, subscriptionId, callbackGasLimit, link, deployerKey) = helperConfig.activeNetworkConfig();
        vm.deal(bob, STARTING_USER_BALANCE);
    }
```
this is my modified setUp function

**Reply by Abhishek-2416** [reply]
I deployed the contract again and now tried to run it 
IT WORKSSS !!1
Thanks brother

**Reply by Abbasjafri-syed** [reply]
ok..it better to run ```forge clean``` if any fix is applied after an error occurs..

**Reply by Abhishek-2416** [reply]
Thanks for your help :)

**Reply by EngrPips** [reply]
Thanks for the help @Abbasjafri-syed

**Reply by Abhishek-2416** [reply]
Although i think its the best method which was suggested by @Abbasjafri-syed , but just found one more method to avoid this error and that is simply run ```forge coverage --ir-minimum``` Even this works but give a warning that it might not be very accurate

**Comment by Abhishek-2416** [comment]
> hey, was getting this error also...finds that Stack too deep error is thrown by the Solidity compiler when any function have lots of local variables, have complex logical expressions or have a deep function stack..
> 
> The main issue is due to using lot of local variable in setup of raffle test which can be mitigated by dividing it into 2 parts like below:
> 
> ```
> 
>     function setUp() public {
>         // setting function for test
>         deployRaffle deploy_raffle = new deployRaffle(); // deploying raffle contract through script
>         (raffle, helper_mode) = deploy_raffle.run(); // assigning contract saved to their type
> 
>         (chainlink_Feed, price, coordinator_vrf, hash,,,,) = helper_mode.active_Config(); //caching value
>         (,,,, call_Limit, confirmations, num_Values, subscriptionId) = helper_mode.active_Config(); // breaking to avoid stack too deep error
>         vm.deal(funder, init_Value); // funding user with amount
>     }
> ```
> 
> try breaking into 2 parts it will solve your compiler issue. i have added my own params as per need but the main thing is break it into 2 parts

**Comment by Abhishek-2416** [comment]
Although i think its the best method which was suggested by @Abbasjafri-syed , but just found one more method to avoid this error and that is simply run forge coverage --ir-minimum Even this works but give a warning that it might not be very accurate

**Comment by aminsammara** [comment]
Adding to the helpful feedback above is you can avoid refactoring your code if you add 
```via-IR = true``` to your `foundry.toml` file at the expense of significant compiling delays (2.5s -> 58s for me)

**Reply by EngrPips** [reply]
Interesting, Thanks for this helpful tip @aminsammara
